clippy_lints/src/arc_new_in_vec_from_slice.rs

```diff
@@ -0,0 +1,82 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::macros::{root_macro_call_first_node, MacroCall};
+use clippy_utils::{last_path_segment, match_def_path, paths};
+use rustc_hir::{Expr, ExprKind, QPath, TyKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `Arc::new` calls in vector initialization using the slice macro `vec![elem; len]`
+    ///
+    /// ### Why is this bad?
+    /// This vector initialization creates `elem` once and clones it `len` times - doing so with `Arc`
+    /// is a bit missleading, because the vector will contain references to the same pointer, although
+    /// it can look like it'll contain different `Arc` instances.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let v = vec![std::sync::Arc::new("some data".to_string()); 100];
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let data = std::sync::Arc::new("some data".to_string());
+    /// let v = vec![data; 100];
+    /// ```
```

Comment:
```suggestion
    /// ### What it does
    /// Checks for `Arc::new` in `vec![elem; len]`
    ///
    /// ### Why is this bad?
    /// This will create `elem` once and clone it `len` times - doing so with `Arc`
    /// is a bit misleading, as it will create references to the same pointer, rather
    /// than different `Arc` instances.
    ///
    /// ### Example
    /// ```rust
    /// let v = vec![std::sync::Arc::new("some data".to_string()); 100];
    /// ```
    /// Use instead:
    /// ```rust
    /// let data = std::sync::Arc::new("some data".to_string());
    /// let v = vec![data; 100];
    /// ```
```

I think this is a little better, correct me if I'm wrong :smile:

Reply:
nah, yours sounds much better 😬